### PR TITLE
Add __radd__ in Circuit

### PR DESCRIPTION
**Pull request by jaeyoo**
This commit adds __radd__ in cirq.Circuit() because Moment + Circuit doesn't work even if Circuit + Moment works.

### CODE REVIEW

File: cirq/circuits/circuit_test.py [hunk 1]
```diff
@@ -247,6 +247,19 @@ def test_add_op_tree():
...
+    ])
+
+    assert cirq.X(a) + c == cirq.Circuit(cirq.X(a))
+    assert [cirq.X(a)] + c == cirq.Circuit(cirq.X(a))
+    assert [[[cirq.X(a)], []]] + c == cirq.Circuit(cirq.X(a))
+    assert (cirq.X(a),) + c == cirq.Circuit(cirq.X(a))
+    assert (cirq.X(a) for _ in range(1)) + c == cirq.Circuit(cirq.X(a))
+    with pytest.raises(AttributeError):
+        _ = cirq.X + c
+
```
[the remark this answers is not in the document]
Comment (jaeyoo) [reply cirq/circuits/circuit_test.py]: Done.

File: cirq/circuits/circuit.py [hunk 2]
```diff
@@ -271,6 +271,9 @@ def __add__(self, other):
         result = self.copy()
         return result.__iadd__(other)
 
+    def __radd__(self, other):
+        return self + other
+
```
[the remark this answers is not in the document]
Comment (jaeyoo) [reply cirq/circuits/circuit.py]: Yes You are right. Thank you for your comment. Please take look at another commit.

File: cirq/circuits/circuit.py [hunk 3]
```diff
@@ -271,6 +271,23 @@ def __add__(self, other):
...
+        device = (self._device if other.device is devices.UNCONSTRAINED_DEVICE
+                  else other.device)
+        device_2 = (other.device if self._device is devices.UNCONSTRAINED_DEVICE
+                    else self._device)
+        if device != device_2:
+            raise ValueError("Can't add circuits with incompatible devices.")
+
+        result = self.copy()
+        return other.__iadd__(result)
```
Comment (c-poole) [inline cirq/circuits/circuit.py]: This will append the circuit to the moment. Copy `other` instead so that the incremental add is performed on the copy instead of the moment or the circuit.
`
result = other.copy()
return result.__iadd__(self)
`

Comment (jaeyoo) [reply cirq/circuits/circuit.py]: Thanks a lot for your kind instructions. It's done now.

**Comment by jaeyoo** [comment]
> LGTM! Thanks for the contribution. We'll need an approval from someone with write access before it can be merged, but it shouldn't take too long.

I really appreciate for helping me ramping up for this first PR! :D

**Comment by dabacon** [comment]
Please add tests for these lines;
```
Line  277 is new or changed but not covered:                  return NotImplemented
Line  286 is new or changed but not covered:              raise ValueError("Can't add circuits with incompatible devices.")
```

**Comment by Strilanc** [comment]
Nit @dabacon : You don't need to request changes that the build system is already enforcing.

**Comment by c-poole** [comment]
> Nit @dabacon : You don't need to request changes that the build system is already enforcing.

Perhaps not, but it is a helpful reminder for a first-time contributor.

**Comment by Strilanc** [comment]
Yeah, the explanation is certainly a good idea.

**Comment by jaeyoo** [comment]
Yes, I need to learn more about github system because I am a novice. Thank you all for your kind comments. I will work on the coverage!

**Comment by jaeyoo** [comment]
I have a question. I realized that __radd__ doesn't require device checking codes inside because (1) __add__ already has it and (2) moment + circuit is converted into cirq.Circuit(moment) + self, which is able to be concatenated with any device. (3) there is no way to set the device of moment, except (4) cirq.Circuit(moment, device=cg.Foxtail) explicitly, for example, then (5) it runs __add__, not __radd__. 

In conclusion, may I just remove the device comparison codes inside __radd__?

**Comment by Strilanc** [comment]
Use backticks \` to avoid the underscores becoming bold `__like_this__` instead of __like_this__.

Yes, if the other instance is a Circuit then `__add__` will be used instead. You can start `__radd__` with a comment `# The Circuit case is handled by __add__`, and then just do an insert at index 0.

**Comment by jaeyoo** [comment]
Thank you, but I am stuck with coverage. I am very confused. Even if I added test codes, coverage still says my codes are not covered. On top of it, it also says my test codes are not covered! Why are my test codes recognized as tests? LOL. I just ran "./check/pytest-and-incremental-coverage" and the messages are:

Running pytest                                                                                                                                 [23/337]
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]                                                                                       
pytest: error: unrecognized arguments: --benchmark-skip                                                                                                
  inifile: None                                                                                                                                        
  rootdir: /usr/local/google/home/jaeyoo/git/Cirq                                                                                                      
                                                                                                                                                       
Finished pytest                                                                                                                                        
Running incremental-coverage                                                                                                                           
************* cirq/circuits/circuit.py (4 uncovered)                                                                                                   
Line  276 is new or changed but not covered:         if not isinstance(other, (ops.Operation, Iterable)):                                              
Line  277 is new or changed but not covered:             return NotImplemented                                                                         
Line  279 is new or changed but not covered:         result = Circuit(other)                                                                           
Line  280 is new or changed but not covered:         return result.__iadd__(self)                                                                      
************* cirq/circuits/circuit_test.py (25 uncovered)                                                                                             
Line  252 is new or changed but not covered:     a = cirq.NamedQubit('a')                                                                              
Line  253 is new or changed but not covered:     b = cirq.NamedQubit('b')                                                                              
Line  255 is new or changed but not covered:     c = cirq.Circuit()                                                                                    
Line  256 is new or changed but not covered:     assert [cirq.X(a), cirq.Y(b)] + c == cirq.Circuit([                                                   
Line  257 is new or changed but not covered:         cirq.Moment([cirq.X(a), cirq.Y(b)]),  
Line  258 is new or changed but not covered:     ])                                                                                             [4/337]
Line  260 is new or changed but not covered:     assert cirq.X(a) + c == cirq.Circuit(cirq.X(a))                                                       
Line  261 is new or changed but not covered:     assert [cirq.X(a)] + c == cirq.Circuit(cirq.X(a))                                                     
Line  262 is new or changed but not covered:     assert [[[cirq.X(a)], []]] + c == cirq.Circuit(cirq.X(a))                                             
Line  263 is new or changed but not covered:     assert (cirq.X(a),) + c == cirq.Circuit(cirq.X(a))                                                    
Line  264 is new or changed but not covered:     assert (cirq.X(a) for _ in range(1)) + c == cirq.Circuit(cirq.X(a))                                  
Line  265 is new or changed but not covered:     with pytest.raises(AttributeError):                                                                  
Line  266 is new or changed but not covered:         _ = cirq.X + c
Line  267 is new or changed but not covered:     with pytest.raises(TypeError):                                                                       
Line  268 is new or changed but not covered:         _ = 0 + c
Line  271 is new or changed but not covered:     d = cirq.Circuit()
Line  272 is new or changed but not covered:     d.append(cirq.Y(b))
Line  273 is new or changed but not covered:     assert [cirq.X(a)] + d == cirq.Circuit([                                                             
Line  274 is new or changed but not covered:         cirq.Moment([cirq.X(a)]),                                                                        
Line  275 is new or changed but not covered:         cirq.Moment([cirq.Y(b)])                                                                         
Line  276 is new or changed but not covered:     ])
Line  277 is new or changed but not covered:     assert cirq.Moment([cirq.X(a)]) + d == cirq.Circuit([                                                
Line  278 is new or changed but not covered:         cirq.Moment([cirq.X(a)]),                                                                        
Line  279 is new or changed but not covered:         cirq.Moment([cirq.Y(b)])                                                                         
Line  280 is new or changed but not covered:     ])

**Comment by jaeyoo** [comment]
> Thank you, but I am stuck with coverage. I am very confused. Even if I added test codes, coverage still says my codes are not covered. On top of it, it also says my test codes are not covered! Why are my test codes recognized as tests? LOL. I just ran "./check/pytest-and-incremental-coverage" and the messages are:
> 
> Running pytest [23/337]
> ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
> pytest: error: unrecognized arguments: --benchmark-skip
> inifile: None
> rootdir: /usr/local/google/home/jaeyoo/git/Cirq
> 
> Finished pytest
> Running incremental-coverage
> ************* cirq/circuits/circuit.py (4 uncovered)
> Line 276 is new or changed but not covered: if not isinstance(other, (ops.Operation, Iterable)):
> Line 277 is new or changed but not covered: return NotImplemented
> Line 279 is new or changed but not covered: result = Circuit(other)
> Line 280 is new or changed but not covered: return result.**iadd**(self)
> ************* cirq/circuits/circuit_test.py (25 uncovered)
> Line 252 is new or changed but not covered: a = cirq.NamedQubit('a')
> Line 253 is new or changed but not covered: b = cirq.NamedQubit('b')
> Line 255 is new or changed but not covered: c = cirq.Circuit()
> Line 256 is new or changed but not covered: assert [cirq.X(a), cirq.Y(b)] + c == cirq.Circuit([
> Line 257 is new or changed but not covered: cirq.Moment([cirq.X(a), cirq.Y(b)]),
> Line 258 is new or changed but not covered: ]) [4/337]
> Line 260 is new or changed but not covered: assert cirq.X(a) + c == cirq.Circuit(cirq.X(a))
> Line 261 is new or changed but not covered: assert [cirq.X(a)] + c == cirq.Circuit(cirq.X(a))
> Line 262 is new or changed but not covered: assert [[[cirq.X(a)], []]] + c == cirq.Circuit(cirq.X(a))
> Line 263 is new or changed but not covered: assert (cirq.X(a),) + c == cirq.Circuit(cirq.X(a))
> Line 264 is new or changed but not covered: assert (cirq.X(a) for _ in range(1)) + c == cirq.Circuit(cirq.X(a))
> Line 265 is new or changed but not covered: with pytest.raises(AttributeError):
> Line 266 is new or changed but not covered: _ = cirq.X + c
> Line 267 is new or changed but not covered: with pytest.raises(TypeError):
> Line 268 is new or changed but not covered: _ = 0 + c
> Line 271 is new or changed but not covered: d = cirq.Circuit()
> Line 272 is new or changed but not covered: d.append(cirq.Y(b))
> Line 273 is new or changed but not covered: assert [cirq.X(a)] + d == cirq.Circuit([
> Line 274 is new or changed but not covered: cirq.Moment([cirq.X(a)]),
> Line 275 is new or changed but not covered: cirq.Moment([cirq.Y(b)])
> Line 276 is new or changed but not covered: ])
> Line 277 is new or changed but not covered: assert cirq.Moment([cirq.X(a)]) + d == cirq.Circuit([
> Line 278 is new or changed but not covered: cirq.Moment([cirq.X(a)]),
> Line 279 is new or changed but not covered: cirq.Moment([cirq.Y(b)])
> Line 280 is new or changed but not covered: ])

Oh, I realized that I didn't install pytest-benchmark in my venv. As well, I missed registering upstream yet. Both make coverage happy & green now :D Thank you for your helps! Now I got it!

**Comment by jaeyoo** [comment]
> Missing coverage for two lines (See my other comment)

I've done. Could you give me LGTM?

**Comment by vtomole** [comment]
Love the simpler implementation.
